### PR TITLE
Improve focus/blur/keyboard events handling

### PR DESCRIPTION
**Pull request by RuslanZavacky**
When the window is focused and unfocused, event triggering works differently. 

1. When the window is unfocused, `el.focus()` is only setting document.activeElement = el;
2. When the window is focused, `el.focus` will set document.activeElement and will also trigger focus event.

This PR will simulate this behavior and help to be consistent between focused and unfocused window.

### CODE REVIEW

**Comment by cibernox** [comment]
@rwjblue @pixelhandler This seems to work better than what we have and I suspect that also better than the focus helper in ember itself (which is private but used inside `click`, `fillIn`, ect...)

I think it worth exploring making a PR to ember itself.

File: addon-test-support/focus.js [hunk 1]
```diff
@@ -22,11 +22,16 @@ export function focus(selector) {
...
           fireEvent(el, 'focusin');
-          fireEvent(el, 'focus', null, false); // focus does not bubble
-        } else {
-          el.focus();
+        }
+
+        el.focus(); // makes `document.activeElement` be `el`. If the browser is focused, it also fires a focus event
+
+        if (browserIsNotFocused) {
+          fireEvent(el, 'focus', null, false); // if the browser is not focused the previous `el.focus()` didn't fire an event, so we simulate it
```
Comment (RuslanZavacky) [inline addon-test-support/focus.js]: Hm, interesting, fireEvent is a bit different interface, null and false is actually never passed correctly there. `buildBasicEvent` has like separate bubbles and cancelable, which are never replaced by the ones passed from outside. I'll look at it a bit later

File: addon-test-support/focus.js [hunk 2]
```diff
@@ -19,14 +19,23 @@ export function focus(selector) {
     let type = el.type;
     if (type !== 'checkbox' && type !== 'radio' && type !== 'hidden') {
```
Comment (RuslanZavacky) [inline addon-test-support/focus.js]: I've just noticed this, that we can't focus on checkbox or radio. That's interesting, as I was not able to find the reason, that we can't focus on them. Can someone point out, what is the reason for that?

Comment (RuslanZavacky) [inline addon-test-support/focus.js]: Based on that http://w3c.github.io/html/editing.html#the-tabindex-attribute it seems, it should be even more elements, that we actually can focus on, so the check, that includes only INPUT, contentEditable and A tag, is a bit too restricted.

File: addon-test-support/-private/is-focusable.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+export default function isFocusable(el) {
```
Comment (cibernox) [inline addon-test-support/-private/is-focusable.js]: Do you need this outside the `focus.js` file? It could just be a function there.

File: addon-test-support/-private/is-focusable.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  let focusableTags = ['INPUT', 'BUTTON', 'LINK', 'SELECT', 'A', 'TEXTAREA'];
+  let { tagName, type } = el;
+
+  if (type === 'hidden') {
+    return false;
+  }
+
+  return focusableTags.includes(tagName) || el.contentEditable;
```
Comment (cibernox) [inline addon-test-support/-private/is-focusable.js]: Avoid includes, since it's not in IE9. Use `indexOf() > -1`

File: addon-test-support/fire-event.js [hunk 5]
```diff
@@ -42,14 +42,12 @@ export function fireEvent(element, type, options = {}) {
...
-  @param {Boolean} (optional) bubbles
-  @param {Boolean} (optional) cancelable
   @return {Event}
   @private
 */
-function buildBasicEvent(type, options = {}, bubbles = true, cancelable = true) {
+function buildBasicEvent(type, options = {}) {
   let event = document.createEvent('Events');
-  event.initEvent(type, bubbles, cancelable);
+  event.initEvent(type, true, true);
```
Comment (cibernox) [inline addon-test-support/fire-event.js]: You must allow to customize if an event is focusable/cancellable from the outside. I think that you cannot do that by merging the options. Perhaps its just an IE thing, but I'm pretty sure this was intentional.

Comment (cibernox) [inline addon-test-support/fire-event.js]: You could however extract `bubbles` and `cancelable` from the options object if that gives you a better API, but I think that you need to use the constructor.

File: addon-test-support/-private/is-focusable.js (hunk 3 above)
Comment (RuslanZavacky) [inline addon-test-support/-private/is-focusable.js]: Same logic is in focus and blur, + if we want to mimic specification, it might be much more logic actually

File: addon-test-support/-private/is-focusable.js (hunk 4 above)
Comment (RuslanZavacky) [inline addon-test-support/-private/is-focusable.js]: hm, babel is not going to polyfill it correctly? I thought its doing it by itself. If not, I'll change back to indexOf

File: addon-test-support/fire-event.js (hunk 5 above)
Comment (RuslanZavacky) [inline addon-test-support/fire-event.js]: I've actually copied that from [ember](https://github.com/emberjs/ember.js/blob/aeeddf7c00b42a340494465252481b5414d87009/packages/ember-testing/lib/events.js#L53) :) Should we change it to be different?

File: addon-test-support/-private/is-focusable.js (hunk 4 above)
Comment (rwjblue) [inline addon-test-support/-private/is-focusable.js]: it will not be polyfilled, we should use `indexOf` 😿

File: addon-test-support/fire-event.js (hunk 5 above)
Comment (cibernox) [inline addon-test-support/fire-event.js]: If you can test it in a IE VM and it works, go for it. I'm surprised, since it's not something I can read in the spec anywhere. The only thing I see in the spec is using the `initEvent` function.

**Comment by RuslanZavacky** [comment]
Interesting behavior with keydown event:

When browser (Chrome) is focused, we get this error:

DOMException: Failed to execute 'createEvent' on 'Document': The provided event type ('KeyEvents') is invalid.

It seems, that KeyEvents is a Firefox thing, not Chrome. Chrome prefers KeyboardEvent. Because of that, keydown is failing when window is not focused :) I am trying to battle it now

**Comment by RuslanZavacky** [comment]
After some trials and failures, I've got to a point, where events are working acceptably fine. It kinda involved quite a lot changes to our tests and app code to adapt correct focusing behaviour. 

## Summary

**focus/blur**
If you want your tests to behave consistently when window is focused and when its not - it involves not only changes to your tests to use focus/blur test helpers, but it also involves changes in your app as well. If in the app you have `.focus()` or `.blur()`, you might have similar behaviour when `.focus` is called and activeElement is changed, but event is not triggered. 

for now, we've ended up with `focus` utility function in the app

```js
export default function focus(element) {
  const documentIsNotFocused = document.hasFocus && !document.hasFocus();

  $(element).focus();

  if (documentIsNotFocused) {
    let event = document.createEvent('Events');
    event.initEvent('focus', false, false);
    element.dispatchEvent(event);
  }
}
```

and instead of doing `el.focus()`, we pass it through this `focus(el)` function

**keyboard events**
This one is quite tricky, as a lot of stuff is deprecated, and other, that is not, not exactly implem ented in other browsers. Implementation where I got to is a bit ugly (with `Object.defineProperty`, but it was closest I was able to get to the consistenly passing tests). 
To create keyboard event, there is kinda a chain of try/catch, as it dependent on if this functionality is working - `new KeyboardEvent()` -> `document.createEvent('KeyboardEvents')` -> `document.createEvent('KeyEvents')` -> `buildBasicEvent(type, options)`. 

The most consistent method was actually building a basic event :)

Comments/suggestions are very much welcome. What you guys think?

File: addon-test-support/fire-event.js [hunk 6]
```diff
@@ -96,13 +94,54 @@ function buildMouseEvent(type, options = {}) {
...
-    let eventOpts = merge(merge({}, DEFAULT_EVENT_OPTIONS), options);
-    event.initKeyEvent(
+    event = new KeyboardEvent(type, eventOpts);
+
+    // Property definitions are required for B/C for keyboard event usage
+    // If this properties are not defined, when listening for key events
+    // keyCode/which will be 0. Also, keyCode and which now are string
+    // and if app compare it with === with integer key definitions,
+    // there will be a fail.
+    Object.defineProperty(event, 'keyCode', {
```
Comment (cibernox) [inline addon-test-support/fire-event.js]: Can you link to some MDN literature about this in the comment.

File: addon-test-support/focus.js [hunk 7]
```diff
@@ -13,23 +14,31 @@ const { run } = Ember;
...
+        fireEvent(el, 'focusin');
+      }
+
+      // makes `document.activeElement` be `el`. If the browser is focused, it also fires a focus event
+      el.focus();
+
+      // if the browser is not focused the previous `el.focus()` didn't fire an event, so we simulate it
+      if (browserIsNotFocused) {
+        fireEvent(el, 'focus', {
+          bubble: false
```
Comment (cibernox) [inline addon-test-support/focus.js]: Should it be **bubbles**, with an **s** at the end?

File: addon-test-support/key-event.js [hunk 8]
```diff
@@ -1,11 +1,16 @@
...
+
+/**
+ * @public
+ * @param selector
+ * @param type
+ * @param keyCode
+ * @param modifiers
+ * @return {*}
+ */
+export function keyEvent(selector, type, keyCode, modifiers = { ctrlKey: false, altKey: false, shiftKey: false, metaKey: false }) {
```
Comment (cibernox) [inline addon-test-support/key-event.js]: I'm 100% for this as before it wasn't possible to simulate `ctrl + A` and things like that. Can you edit the readme and add docs for this? Perhaps it's easier if you improve `keyEvent` in a separate PR.

**Comment by cibernox** [comment]
I left a couple questions, but looks good to me otherwise.

File: addon-test-support/key-event.js (hunk 8 above)
Comment (RuslanZavacky) [inline addon-test-support/key-event.js]: edited :)

**Comment by cibernox** [comment]
LGTM.

**Comment by RuslanZavacky** [comment]
🎉

**Comment by cibernox** [comment]
I'll release 0.4.0 since changes are somewhat significative. Waiting for another PR.

**Comment by john-kurkowski** [comment]
🎉 

Yeah, take care in releasing this. Sounds like it diverges from (but improves upon 🤞) Ember's own test helpers **and** changes client usage.

**Comment by cibernox** [comment]
@john-kurkowski The public API should be the same, so I think that this should work in every case the test helpers in Ember work, and new cases where they didn't work.

**Comment by john-kurkowski** [comment]
Oops. Maybe I was reading too much into [Ruslan's comment](https://github.com/cibernox/ember-native-dom-helpers/pull/62#issuecomment-298616533).

> If you want your tests to behave consistently when window is focused and when its not - it involves not only changes to your tests to use focus/blur test helpers, but it also involves changes in your app as well.

We've been dogfooding this PR. We had to make changes to our test suite. Maybe it's only if the client cares about these focus/blur/background edge cases. Maybe it will be a drop-in upgrade for most clients.

**Comment by cibernox** [comment]
@john-kurkowski As I understand it, you shouldn't need to make changes to your app if you run your tests with the browser's window focused. On that situation it works as it did before.
You might need to make changes to your app to make tests pass when the browser doesn't have the focus, but that didn't work before either.

@RuslanZavacky Am I right?

**Comment by RuslanZavacky** [comment]
@cibernox yeap, exactly

**Comment by pixelhandler** [comment]
@john-kurkowski the part about `changes in your app` is for better and consistent test behavior, not a breaking change. if you want the app to know it can be run on a browser w/o focus then the focus helper is nice to have in the app.
